### PR TITLE
feat: Added searchable files list for IDE tabs

### LINKED ISSUE #32479: [Task] - IDE tabs revamp - searchable menu

### PR DESCRIPTION
**Pull request by albinAppsmith**
Fixes #32479

## Automation

/ok-to-test tags=""

### :mag: Cypress test results

> [!CAUTION]  
> If you modify the content in this section, you are likely to disrupt the CI result for your PR.

### CODE REVIEW

File: app/client/src/ce/constants/messages.ts [hunk 1]
```diff
@@ -2467,3 +2467,11 @@ export const SPLITPANE_ANNOUNCEMENT = {
...
 export const CANVAS_VIEW_MODE_TOOLTIP = (shortcutKey: string) =>
   `💡 ${shortcutKey} click a widget to navigate to UI mode.`;
+
+// searchable files list
+export const SFL = {
+  GROUP_LABEL: (type: string) => `Open a ${type}`,
+  OPENED_GROUP_LABEL: () => `Opened`,
+  QUERY_TEXT: `query`,
+  JS_OBJECT_TEXT: `JS object`,
+};
```
[the remark this answers is not in the document]
Comment (mohanarpit) [reply app/client/src/ce/constants/messages.ts:2477]: @albinAppsmith Can you help me understand why we have resolved this suggestion without making the code changes?

[the remark this answers is not in the document]
Comment (albinAppsmith) [reply app/client/src/ce/constants/messages.ts:2477]: @mohanarpit I have not made this change because that variable name will be longer and I didn't see any particular advantage to use that long name. That's why I have added the comment in the messages file. The place where I use this variable, the file name is already called `SearchableFilesList`.

[the remark this answers is not in the document]
Comment (mohanarpit) [reply app/client/src/ce/constants/messages.ts:2477]: Hmm ... OK. I'm not completely onboard but I get your POV. 👍🏽

File: app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    // reset filter
+    setFiles(allItems);
+    setTabs(openTabs);
+  }, [isOpen]);
+
+  if (allItems.length === 0) {
+    return null;
+  }
+
+  const filterHandler = (value: string) => {
```
Comment (hetunandu) [inline app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx:53]: Please use Fuse.js for this. It is already a dependency and allows us to do fuzzy search as well

File: app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+          <ListIconContainer>{file.icon}</ListIconContainer>
+          <TabTextContainer>{file.title}</TabTextContainer>
+        </Flex>
+      </MenuItem>
+    ));
+
+  return (
+    <Menu onOpenChange={setOpen} open={isOpen}>
+      <MenuTrigger>
+        <StyledButton
```
Comment (hetunandu) [inline app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx:76]: This should be a togglebutton

**Review by hetunandu — CHANGES_REQUESTED** [review]

File: app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
+import React, { useEffect, useState } from "react";
```
Comment (hetunandu) [inline app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx:1]: Lets add a unit test for this component as well

**Review by hetunandu — CHANGES_REQUESTED** [review]

File: app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx (hunk 3 above)
Comment (albinAppsmith) [reply app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx:76]: In design it is not a toggle button, and ideally menu trigger should not be a toggle button.

File: app/client/src/ce/constants/messages.ts [hunk 5]
```diff
@@ -2467,3 +2467,11 @@ export const SPLITPANE_ANNOUNCEMENT = {
 
 export const CANVAS_VIEW_MODE_TOOLTIP = (shortcutKey: string) =>
   `💡 ${shortcutKey} click a widget to navigate to UI mode.`;
+
+// searchable files list
+export const SCROLLABLE_FILES_LIST = {
```
Comment (hetunandu) [inline app/client/src/ce/constants/messages.ts:2472]: I thought this was `SEARCHABLE_FILES_LIST`

File: app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.test.tsx [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    // Check if navigateToTabMock was called with the correct item
+    expect(navigateToTabMock).toHaveBeenCalledWith(openTabs[0]);
+
+    // Click on a open a js/query group menu item
+    fireEvent.click(allByText[1]);
+
+    // Check if navigateToTabMock was called with the correct item
+    expect(navigateToTabMock).toHaveBeenCalledWith(allItems[0]);
+  });
```
[the remark this answers is not in the document]
Comment (hetunandu) [reply app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.test.tsx:84]: @albinAppsmith this suggestion makes sense

File: app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx (hunk 3 above)
Comment (hetunandu) [reply app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx:87]: It may be missed in the designs, the menu does behave as if it is toggleble. We can confirm with @momcilo-appsmith once

Comment (albinAppsmith) [reply app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.tsx:87]: I would say menu should not ever triggered from a toggle button, it is not a standard behaviour. Still will confirm once.

File: app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.test.tsx [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import {
+  EditorEntityTab,
+  type EntityItem,
+} from "@appsmith/entities/IDE/constants";
+import { PluginType } from "entities/Action";
+
+// Mock the useCurrentEditorState hook
+jest.mock("../hooks", () => ({
+  useCurrentEditorState: jest.fn(() => ({ segment: EditorEntityTab.JS })),
+}));
```
[the remark this answers is not in the document]
Comment (albinAppsmith) [reply app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.test.tsx:13]: Will add more test once Mončilo refines the empty states and other behaviours.

File: app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.test.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (albinAppsmith) [reply app/client/src/pages/Editor/IDE/EditorTabs/SearchableFilesList.test.tsx]: Made this change.
https://github.com/coderabbitai

**Review by hetunandu — APPROVED** [review]
